src/core/api_product/nodo_pagamenti_api/decoupler/decoupler-algorithm.xml

```diff
@@ -331,9 +333,21 @@ new JProperty("ttl", 1)
                             isWhitelisted = (brokersWhitelisted || brokers.Contains(pspBrokerId)) &&
                                             (channelsWhitelisted || channels.Contains(channelId)) &&
                                             isCiAndStationOk;
+                            // check if 'tipoVersamento' is accepted for WISP dismantling only
+                            // if is whitelisted and it is a 'nodoInviaRPT'
+                            bool isNodoInviaRPT = "nodoInviaRPT".Equals((string)context.Variables["soapAction"]);
+                            if (isWhitelisted && isNodoInviaRPT) {
+                                byte[] rptAsBytearray = Convert.FromBase64String((string) primitive.Descendants("rpt").FirstOrDefault());
+                                string rptAsString = System.Text.Encoding.UTF8.GetString(rptAsBytearray);
+                                XElement rpt = XElement.Parse(rptAsString);
+                                string rptPaymentType = rpt.Descendants(rpt.Name.Namespace + "tipoVersamento").FirstOrDefault()?.Value;
+                                string[] paymentTypes =  ((string) context.Variables["wisp_whitelisted_nodoinviarpt_paymenttypes"]).Split(',');
+                                isWhitelisted = paymentTypes.Contains(rptPaymentType);
+                            }
```

Comment:
- shift up this code before `isWhitelisted = (brokersWhitelisted ...`
- set vars type `isValidNodoInviaRptOrNodoInviaCarrelloRPT` with something like that 
`isValidNodoInviaRptOrNodoInviaCarrelloRPT` = `isNodoInviaCarrelloRPT` or `isNodoInviaRptWithAllowedPaymentType`

Reply:
Yes, it's a good idea. I will apply your suggestions on the code. 🔝

Reply:
I have updated the code following your suggestion. Is it correct and is it what you meant in your comment?